### PR TITLE
initial commit of Cube to be used in Metrics Store

### PR DESCRIPTION
**Pull request by abaranau**
For https://issues.cask.co/browse/CDAP-1405 (Implement basic Cube dataset for storing metrics data), part of https://issues.cask.co/browse/CDAP-1404 (Refactor Metrics Storage layer) which is part of https://issues.cask.co/browse/CDAP-1138 (Metrics Operations Dashboard V2)

### CODE REVIEW

**Comment by abaranau** [comment]
Build is GREEN: https://builds.cask.co/browse/CDAP-DUT702

File: cdap-watchdog/src/test/java/co/cask/cdap/metrics/store/cube/AbstractCubeTest.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+  private List<TimeValue> timeValues(long... longs) {
+    List<TimeValue> timeValues = Lists.newArrayList();
+    for (int i = 0; i < longs.length; i += 2) {
+      timeValues.add(new TimeValue(longs[i], longs[i + 1]));
+    }
+    return timeValues;
+  }
+
+  private Map<String, String> tagValuesWithNulls(String... tags) {
```
Comment (gsps1) [inline cdap-watchdog/src/test/java/co/cask/cdap/metrics/store/cube/AbstractCubeTest.java]: method name is not relevant ? i don't see null values passed for this ?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/DefaultCube.java [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+      for (String tagName : query.getGroupByTags()) {
+        // todo: use Map<String, String> instead of List<TagValue> into a String, String, everywhere
+        for (TagValue tagValue : next.getTagValues()) {
+          if (tagName.equals(tagValue.getTagName())) {
+            if (tagValue.getValue() == null) {
+              // Currently, we do NOT return null as grouped by value.
+              // Depending on whether tag is required or not the records with null value in it may or may not
+              // aggregation. At this moment, the choosing of the aggregation for query doesn't look at this,
+              // potentially null may or may not be included in results, depending on the aggregation selected
+              // querying. We don't want to produce inconsistent results, so don't return nulls in any of the
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/DefaultCube.java]: comment is not complete ?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/DefaultAggregation.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ * Default implementation of {@link Aggregation}.
+ */
+public class DefaultAggregation implements Aggregation {
+  private final List<String> aggregateTags;
+  private final Set<String> requiredTags;
+
+  /**
+   * Creates instance of {@link DefaultAggregation}.
+   * @param aggregateTags tags to be included in aggregation.
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/DefaultAggregation.java]: can be good to point out here that the order of tags is important ?

**Comment by gsps1** [comment]
just got a couple of minor questions, LGTM otherwise.

File: cdap-watchdog/src/test/java/co/cask/cdap/metrics/store/cube/AbstractCubeTest.java (hunk 1 above)
Comment (abaranau) [inline cdap-watchdog/src/test/java/co/cask/cdap/metrics/store/cube/AbstractCubeTest.java]: right. Fixed

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/DefaultCube.java (hunk 2 above)
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/DefaultCube.java]: oops bad merge

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/DefaultAggregation.java (hunk 3 above)
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/DefaultAggregation.java]: very good point. Added to Aggregation javadoc and added link from here to it.

**Pull request merged**